mos-platform/geos-cbm/README.md

```diff
@@ -0,0 +1,415 @@
+
+Using this `#pragma` all functions and variables are stored in the VLIR chain 01. The only exception are string constants. There is a string pooling in place. By default all strings are stored in the resident part (VLIR chain 0). To come around this, please declare initialized static variables to hold the string value:
+
+    /* Define string constants as static variables. This way they will be stored
+     * in the vlir record 2 and not in the main constant string pool.
+     */
+    char cut_msg[] = "cut handler called.";
+    char copy_msg[] = "copy handler called.";
+    char paste_msg[] = "paste handler called.";
+    char icon1_msg[] = "icon handler called.";
+
+This way the string constant will be stored in the VLIR chain set by the `#pragma clang section` command.
+
+## Swapping in code for an overlay
+
+While the linker will store the artifacts for an overlay in a VLIR chain, it is on you to load that code and static data in when needed. First you need to stored the track/sector pointers for each VLIR chain in a swap table at the startup of the application. Please note that the GEOS file type given in the FindFTypes call must match the GEOS file type  of your application:
+
+	/* holds (T,S) pairs, one for each module of our application */
+	tr_se_pair_t  swapTable[NUM_MODS];    		
+	
+	/* holds currently swapped in module number */
+	uint8_t  curModule;
+	
+	/*********************************************************************
+    * This routine sets up a table which contains the track and sector
+    * numbers for each of the program modules which can be loaded.
+    * This table will be used by the SwapMod function later on.
+    *********************************************************************/
+    void InitSwap(void)
+    {
+    	char appName[1][17];
+    	
+    	FindFTypes(appName, APPLICATION, 1, "Sample VLIR V1.0");
+    	OpenRecordFile(appName[0]);
+    	MoveData((const uint8_t*)&fileHeader+4, (uint8_t*)swapTable, NUM_MODS*2);
+    	CloseRecordFile();
+    	curModule = 0xff;
+    }
+
+Whenever you now need a function or static data of the overlay, you need to load in that VLIR chain BEFORE you use that artifacts:
+
+    /*********************************************************************
+    * This routine swaps a module in. Note how it used "ReadFile" instead
+    * of "ReadRecord" so that it does not affect any opened VLIR file.
+    *********************************************************************/
+    noinline void SwapMod(uint8_t req_mod_no)
+    {
+	    if (req_mod_no != curModule) {
+		    curModule = req_mod_no;
+		    ReadFile(OVERLAYADDR, OVERLAYSIZE, swapTable[req_mod_no-1]);
+	    }
+    }
+
+`OVERLAYADDR` and `OVERLAYSIZE` are defined by the linker and default to `0x5000` and `0x1000` (for details see the `vlir.ld` linker script in `.../mos-platform/geos-cbm/vlir.ld`).
+You can change the size of the overlay memory to e.g. 2k bytes by adding the following statement to any of your C files:
```

Comment:
0x2000 isn't 2KiB; it's 8KiB. (I do this a lot ;) The thing to remember is that 0x1000 is 4KiB.